### ISSUE TITLE
Processing Linq expressions may cause exceptions

### ISSUE THREAD

**Issue by vsoldatkin** [esskar/Serialize.Linq#41]
Somehow, I experience a couple of tests failing.

It occures when a Linq expression is processed:
```
Expression<Func<IEnumerable<int>, IEnumerable<int>>> interferingExpression = 
c =>  from x in c
         let anotherConst = 100
         where (x == anotherConst)
         select x;
```
"Where" predicate is pushed into an anonymous class by a compiler, and finally ExpressionContext.ResolveType may find a wrong anonymous class (from another assembly for example) and fail to procceed further due to the mismatch.

**Comment by vsoldatkin** [comment]
I tried to provide a couple of reliable test cases for the issue.
Please could you confirm that the issue occures on these cases in your environment as well?

**Comment by esskar** [comment]
it is actually strange.
for me, it finds the declaring type that is somehow defined here
https://github.com/esskar/Serialize.Linq/blob/master/src/Serialize.Linq.Tests/Internals/SerializerTestData.cs#L54

but that looks wrong because the type is retieved from an expression member and is not guessed or similar

**Comment by vsoldatkin** [comment]
Exactly. 
It seems this linq expression requires an f__AnonymousType0``2, which is essentially a 
```
new 
{
     x,
     anotherConst
}
```
and is probably used in "where" comparison.

The problem is there may be a lot of similar types (f__AnonymousType0``2, f__AnonymousType1``2, etc) in other assemblies. So a naming collision is very probable.

The collision takes effect in the loop https://github.com/esskar/Serialize.Linq/blob/master/src/Serialize.Linq/ExpressionContext.cs#L52 It does not take into account the origin of a desired type.

Searching types by an assembly qualified name may be a solution.

**Comment by esskar** [comment]
I think i fixed it with this c18f90333905eb02b00881d005668b27a9adabac commit.
Can you have a look and retest your stuff please?

**Comment by vsoldatkin** [comment]
Seems ok to me.
Thank you.

**Comment by esskar** [comment]
I just published version 1.2.2. I owe you a beer!
